enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,893 @@
+
+1. The upgrade infrastructure of the cluster ensures that all the images
+required for the upgrade are pinned and pre-loaded in all the nodes of the
+cluster. These images will be un-pinned during the next upgrade.
+
+In addition, for the cases were the cluster is completely disconnected or it
+isn't possible to use a registry server:
+
+1. An engineer uses the `oc adm upgrade create bundle ...` tool described in
+this enhancement to prepare the upgrade bundle containing all the artifacts
+(container images and metadata) that are required to perform the upgrade plus
+the images required for the custom workloads, and writes it to a USB stick (or
+any other suitable media) that will be handed over to the technicians
+responsible for performing the upgrades, together with documentation explaining
+how to use it.
+
+1. The technicians receive copies of the USB stick and the corresponding
+documentation.
+
+1. The technician goes to the cluster site and uses the upgrade bundle inside
+the to perform the upgrade. The documentation will ask the technician to
+plug the USB stick in one of the nodes of the cluster and provide simple
+instructions to verify that the upgraded has been applied correctly. This step
+is potentially repeated multiple times by the same technician for multiple
+clusters using the same USB stick or copies of it.
+
+Note that the upgrade bundle should not be specific for a particular cluster,
+only for the OpenShift architecture and version. Technicians should be able to
+use that package for any cluster with that architecture and version.
+
+### API Extensions
+
+There are no new object kinds introduced by this enhancement, but new fields
+will be added to existing `ClusterVersion` and `ContainerRuntimeConfig` objects.
+
+The new fields for the `ClusterVersion` object are defined in detail in the
+in https://github.com/openshift/api/pull/1548.
+
+The new fields for the `ContainerRuntimeConfig` object are defined in detail in
+https://github.com/openshift/machine-config-operator/pull/3839.
+
+### Implementation Details/Notes/Constraints
+
+The proposed solution is based on pre-loading and pinning all required images
+in all the nodes of the cluster before starting the upgrade, and ensuring that
+no component requires access to a registry server during the upgrade. For this
+to work the following changes are required:
+
+1. No OpenShift component used during the upgrade should use the `Always` pull
+policy, as that forces the kubelet and CRI-O to try to contact the registry
+server even if the image is already available.
+
+1. No OpenShift component should garbage collect the images required for the
+upgrade. This is typically started by the kubelet instructing CRI-O to remove
+images.
+
+1. No OpenShift component should explicitly try to contact the registry server
+without a fallback alternative.
+
+1. The machine config operator needs to support image pinning, pre-loading and
```

Comment:
Do we perform image pinning only for OCP specific images or it also includes customer workload images?

Reply:
It depends. But in high-level, the right approach would be:
1. Pin OCP images - for cases where the cluster is used to run customer workloads only
2. Pin OCP images and partner images - for cases where partners have an appliance (basically OCP + more infra workloads they need to run) - in this case, you'll want to have both OCP and partner images pinned. While the customer workloads aren't.

An example, Dell ships an appliance that includes OCP + powerflex storage. You'd need both OCP images and powerflex related images pinned. This setup is used by customers to run their workloads. They can host these images in a registry (in cluster) and pull those from there. But the images required for OCP + Dell software + the in-cluster registry, should be pinned so that everything is running (for reboot, upgrade, etc.).

Reply:
ah got it, thanks for the explanation.